File: 139_Word_Break/C/Solution.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   Solution.c                                         :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: inazaria <inazaria@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/06/21 10:36:45 by inazaria          #+#    #+#             */
/*   Updated: 2024/06/21 17:11:51 by inazaria         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>

bool	isValidWord(char *s, size_t index_s, char *word) {
	if (strncmp(s + index_s, word, strlen(word)) == 0)
		return true;

	return false;
}

// dp[x] == 0 -> not visited
// dp[x] == 1 -> visited and true
// dp[x] == 2 -> visited and false

bool	backtracking(char *s, size_t index_s, char **wordDict, int wordDictSize, int *dp, size_t size_dp) {
	if (index_s >= strlen(s))
		return true;

	if (dp[index_s] > 0)
		return (dp[index_s] == 1) ? true : false;

	for (int i = 0; i < wordDictSize; i++) {
		if (isValidWord(s, index_s, wordDict[i]) == true) {
			if (backtracking(s, index_s + strlen(wordDict[i]), wordDict, wordDictSize, dp, size_dp) == true) {
				dp[index_s] = 1;
				return true;
			}
		}
	}

	dp[index_s] = 2;
	return false;
}

bool	wordBreak(char* s, char** wordDict, int wordDictSize) {	

	int dp[strlen(s)];
	bzero(dp, strlen(s) * sizeof(int));

	return backtracking(s, 0, wordDict, wordDictSize, dp, 0);
}


int main(int argc, char *argv[]) {

	if (argc < 2) {
		return (1);
	}

	int	wordDictSize = argc - 2;
	char *wordDict[wordDictSize];

	for (int i = 2; i < argc; i++)
		wordDict[i - 2] = argv[i];


	bool answer = wordBreak(argv[1], wordDict, wordDictSize);
	(answer) ? printf("answer is True\n") : printf("Asnwer is False\n");
	return (0);
}

File: 139_Word_Break/python3/Solution.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    Solution.py                                        :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: inazaria <inazaria@student.42.fr>          +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/06/17 01:43:05 by inazaria          #+#    #+#              #
#    Updated: 2024/06/21 22:46:39 by inazaria         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

import sys
from typing import List

class Solution:

    def isValidWord(self, s, index_s, word) -> bool :
        len_word = len(word)
        return s[index_s:index_s + len_word] == word


    def backtrack(self, s: str, index_s: int, wordDict: List[str], dp: dict[int, bool]) -> bool :
        
        if index_s in dp :
            return dp[index_s]

        if index_s >= len(s) :
            return True

        for word in wordDict :
            if self.isValidWord(s, index_s, word) :
                dp[index_s] = True
                if self.backtrack(s, index_s + len(word), wordDict, dp) :
                    return True
                
        dp[index_s] = False

        return False

    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        if s in wordDict : 
            return True

        dp: dict[int, bool] = {}
        return self.backtrack(s, 0, wordDict, dp)


if __name__ == "__main__" :
    if (len(sys.argv) != 3) :
        print("Usage : python3 Solution.py <string> <wordDict>")
        print("Example : python3 Solution \"catsandog\" \"[\"cats\",\"dog\",\"sand\",\"and\",\"cat\"]")
        sys.exit(1)

    wordDict = []
    string = ""
    try : 
        string: str = str(sys.argv[1])
        wordDict: list[str] = eval(sys.argv[2])
    except :
        print("Error : Bad Formatting")
        print("Usage : python3 Solution.py <string> <wordDict>")
        print("Example : python3 Solution \"catsandog\" \"[\"cats\",\"dog\",\"sand\",\"and\",\"cat\"]")
    
    answer: bool = Solution().wordBreak(string, wordDict)
    print(f"string = {string}, wordDict = {wordDict}, answer = {answer}")
